src/memory_manager.rs

```diff
@@ -287,13 +439,103 @@ impl<M: Memory> MemoryManagerInner<M> {
             bucket_size_in_pages: header.bucket_size_in_pages,
             memory_sizes_in_pages: header.memory_sizes_in_pages,
             memory_buckets,
+            freed_buckets,
+            first_unused_bucket,
+        }
+    }
+
+    fn load_layout_v2(memory: M, header: Header) -> Self {
+        const BYTE_SIZE_IN_BITS: usize = 8;
+        const BUCKETS_INDEX_SIZE_IN_BYTES: usize =
+            (MAX_NUM_BUCKETS as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
+                / BYTE_SIZE_IN_BITS;
+        let mut buckets = vec![0; BUCKETS_INDEX_SIZE_IN_BYTES];
+        memory.read(bucket_indexes_offset().get(), &mut buckets);
+
+        let buckets_decompressed = bytes_to_bucket_indexes(&buckets);
+
+        let mut unallocated_buckets: BTreeSet<u16> = (0..MAX_NUM_BUCKETS as u16).collect();
+        let mut memory_buckets = BTreeMap::new();
+
+        let mut last_occupied_bucket: i32 = -1;
+
+        let mut j = 0;
+        for (memory, memory_size_in_pages) in header.memory_sizes_in_pages.into_iter().enumerate() {
+            let memory_size_in_buckets =
+                (memory_size_in_pages + header.bucket_size_in_pages as u64 - 1)
+                    / header.bucket_size_in_pages as u64;
+            let mut vec_buckets = vec![];
+            for _ in 0..memory_size_in_buckets {
+                let bucket = buckets_decompressed[j];
+                last_occupied_bucket = std::cmp::max(bucket.0 as i32, last_occupied_bucket);
+                vec_buckets.push(bucket);
+                unallocated_buckets.remove(&bucket.0);
+                j += 1;
+            }
+            memory_buckets
+                .entry(MemoryId(memory as u8))
+                .or_insert(vec_buckets);
+        }
+
+        let mem_mgr = if last_occupied_bucket == -1 {
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets: LinkedList::new(),
+                first_unused_bucket: BucketId(0),
+            }
+        } else {
+            let first_unused_bucket = BucketId((last_occupied_bucket + 1) as u16);
+
+            let mut freed_buckets = LinkedList::new();
+            for i in unallocated_buckets.iter() {
+                if *i < first_unused_bucket.0 {
+                    freed_buckets.push_back(BucketId(*i));
+                }
+            }
```

Comment:
Aren't we here still storing all the unallocated buckets up to `MAX_NUM_BUCKETS`? Per my comment yesterday, that's not necessary and it's more efficient to store only the buckets that were previously freed (i.e. buckets with ids <= last_occupied_bucket)

Reply:
Yes, that would be preferable, because in the future we could increase the number of buckets, and in that case it's not desirable to allocate a collection that's `MAX_NUM_BUCKETS` in size.

Reply:
Done.